File: core/core-domain/src/main/java/com/coffee/gu/stamp/StampRevertPlan.java
package com.coffee.gu.stamp;

import com.coffee.gu.coupon.IssuedCoupon;

import java.util.List;

public class StampRevertPlan {

    private final List<Stamp> stampsToCancel;
    private final List<Stamp> usedStamps;
    private final List<IssuedCoupon> rewardCouponsToCancel;

    public StampRevertPlan(
            List<Stamp> stampsToCancel,
            List<Stamp> usedStamps,
            List<IssuedCoupon> rewardCouponsToCancel
    ) {
        this.stampsToCancel = stampsToCancel;
        this.usedStamps = usedStamps;
        this.rewardCouponsToCancel = rewardCouponsToCancel;
    }

    public static StampRevertPlan empty() {
        return new StampRevertPlan(List.of(), List.of(), List.of());
    }

    public boolean isEmpty() {
        return stampsToCancel.isEmpty();
    }

    public boolean hasRewardCouponsToCancel() {
        return !rewardCouponsToCancel.isEmpty();
    }

    public boolean hasUsedRewardCoupon() {
        return rewardCouponsToCancel.stream()
                .anyMatch(IssuedCoupon::isUsed);
    }

    public List<Stamp> getStampsToCancel() {
        return stampsToCancel;
    }

    public List<IssuedCoupon> getRewardCouponsToCancel() {
        return rewardCouponsToCancel;
    }

    public int cancelStampCount() {
        return stampsToCancel.size();
    }
}

File: core/core-domain/src/main/kotlin/com/coffee/gu/stamp/StampRevertPlan.kt
package com.coffee.gu.stamp

import com.coffee.gu.coupon.IssuedCoupon

class StampRevertPlan(
    val stampsToCancel: List<Stamp>,
    val usedStamps: List<Stamp>,
    val rewardCouponsToCancel: List<IssuedCoupon>,
) {
    val isEmpty: Boolean
        get() = stampsToCancel.isEmpty()

    val hasRewardCouponsToCancel: Boolean
        get() = rewardCouponsToCancel.isNotEmpty()

    val hasUsedRewardCoupon: Boolean
        get() = rewardCouponsToCancel.any { it.isUsed }

    fun hasRewardCouponsToCancel(): Boolean = hasRewardCouponsToCancel
    fun hasUsedRewardCoupon(): Boolean = hasUsedRewardCoupon

    fun cancelStampCount(): Int {
        return stampsToCancel.size
    }

    companion object {
        fun empty(): StampRevertPlan {
            return StampRevertPlan(
                stampsToCancel = emptyList(),
                usedStamps = emptyList(),
                rewardCouponsToCancel = emptyList(),
            )
        }
    }
}
